var/spack/repos/builtin/packages/xyce/package.py

```diff
@@ -0,0 +1,82 @@
+
+from spack import *
+
+
+class Xyce(CMakePackage):
+    """Xyce (zīs, rhymes with “spice”) is an open source, SPICE-compatible,
+    high-performance analog circuit simulator, capable of solving extremely
+    large circuit problems by supporting large-scale parallel computing
+    platforms. It also supports serial execution on all common desktop platforms,
+    and small-scale parallel runs on Unix-like systems. In addition to analog
+    electronic simulation, Xyce has also been used to investigate more general
+    network systems, such as neural networks and power grids.
+    """
+
+    homepage = 'https://xyce.sandia.gov'
+    git      = 'https://github.com/Xyce/Xyce.git'
+    url      = 'https://github.com/Xyce/Xyce/archive/Release-7.2.0.tar.gz'
+    maintainers = ['kuberry']
+
+    version('master',  branch='master')
+    version('7.2.0', 'cf49705278ecda46373784bb24925cb97f9017b6adff49e4416de146bdd6a4b5', preferred=True)
+
+    depends_on('cmake@3.13:', type='build')
+    depends_on('flex')
+    depends_on('bison')
+
+    variant('build_type', default='Release',
+            description='CMake build type',
+            values=('Debug', 'Release', 'RelWithDebInfo', 'MinSizeRel'))
+
+    variant('mpi', default=True, description='Enable MPI support')
+    depends_on('mpi', when='+mpi')
+
+    variant('pymi', default=False, description='Enable Python Model Interpreter for Xyce')
+    depends_on('python@3:', type=('build', 'link', 'run'), when='+pymi')
+    depends_on('py-pip', type='run', when='+pymi')
+    depends_on('py-pybind11@2.6.1:', when='+pymi')
+
+    depends_on('trilinos@12.12.1~adios2~alloptpkgs+amesos+amesos2+anasazi+aztec+belos~boost~cgns~chaco+complex~cuda~cuda_rdc~debug~dtk+epetra+epetraext~exodus+explicit_template_instantiation~float+fortran~glm~gtest+hdf5~hwloc~hypre+ifpack~ifpack2~intrepid~intrepid2~ipo+isorropia+kokkos~matio~mesquite~metis~minitensor~ml+mpi~muelu~mumps~netcdf+nox~openmp~phalanx~piro~pnetcdf~python~rol~rythmos+sacado~shards~shared~shylu~stk~stratimikos~strumpack+suite-sparse~superlu~superlu-dist~teko~tempus+teuchos+tpetra+trilinoscouplings~wrapper~x11~xsdkflags~zlib+zoltan~zoltan2+stokhos+amesos2basker+epetraextbtf+epetraextexperimental+epetraextgraphreorderings gotype=\'none\'', when="+mpi")
+    depends_on('trilinos@12.12.1~adios2~alloptpkgs+amesos+amesos2+anasazi+aztec+belos~boost~cgns~chaco+complex~cuda~cuda_rdc~debug~dtk+epetra+epetraext~exodus+explicit_template_instantiation~float+fortran~glm~gtest+hdf5~hwloc~hypre+ifpack~ifpack2~intrepid~intrepid2~ipo+isorropia+kokkos~matio~mesquite~metis~minitensor~ml~mpi~muelu~mumps~netcdf+nox~openmp~phalanx~piro~pnetcdf~python~rol~rythmos+sacado~shards~shared~shylu~stk~stratimikos~strumpack+suite-sparse~superlu~superlu-dist~teko~tempus+teuchos+tpetra+trilinoscouplings~wrapper~x11~xsdkflags~zlib+zoltan~zoltan2+stokhos+amesos2basker+epetraextbtf+epetraextexperimental+epetraextgraphreorderings gotype=\'none\'', when="~mpi")
```

Comment:
Why are we getting _this_ specific? Would this package not build if I use e.g. `trilinos+adios2`?

Reply:
Unfortunately, it is necessary to be this specific at this time. A little background:
- Trilinos has a very large number of packages available, most of which would not be built by default in CMake
     - This means enabling any given package will require a variant
- For the users of the Trilinos recipe to-date (xSDK, Deal.ii, DTK), they have some packages that they like to enable by default, but these are not the packages that we would enable
    - This means unless we explicitly disable certain variants, then we will build many things we don't need

There are future plans (Fall), to move towards a Trilinos recipe that will disable all packages by default, and then users will explicitly enable the ones they want (like in CMake). However, even that will not result in a much less specific recipe (it will just remove all of the ~variants).

Reply:
> This means unless we explicitly disable certain variants, then we will build many things we don't need

This is perfectly fine, at a package level, since the package is a place where requirements (not preferences) are stated. I understand that spending time to build unnecessary things is disappointing, but you can avoid that by customizing defaults in `packages.yaml`. Think of it the other way around: people that already installed e.g. trilinos with everything will need to install it again to match the required absence of some components here.

TL;DR can we remove already the `~variant`?

Reply:
I can definitely appreciate what your are saying. And I think for most packages this would be the correct way to go. However, in this particularly case, Xyce requires the user to use version 12.12.1 of Trilinos, which is now several years old. There is a near 0% likelihood that the user will already have this version of Trilinos installed on their machine, unless it was specifically built to be used with Xyce.

Reply:
Is there a way to disable the variant unless the user has **already** built that package with the variant? Basically, if we don't put ~variant, then it may be built and we don't need it. However, if it is already built, then we can use it (I don't think this exists in Spack, but I'm also new to Spack).

Also, Trilinos will update their recipe later this year to remove the needed for ~variants, but that is not yet the case.

Reply:
I've added that comment to the PR. Thanks for the suggestion to add the explanation.